### PR TITLE
Use gRPC for distributor <-> ingester rpcs.

### LINKED ISSUE #105: Switch to GRPC for distributor <-> ingester rpc
Should remove some boilerplate and hopefully reduce latency.

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #105

- Don't commit generated code, teach make how to build it
- Don't duplicate remote.proto from promtheus, import it.
- Make ingester and distributor expose grpc services
- Move http server code into the ingester/distributor packages (some duplication, but will go away after we migrate)
- Ingesters still expose http interface for migration purposes.

TODO:
- [x] Instrument grpc https://github.com/grpc-ecosystem/go-grpc-prometheus

### CODE REVIEW

**Comment by rade** [comment]
From #105:
> Should remove some boilerplate and hopefully reduce latency.

Any measurements?

**Comment by tomwilkie** [comment]
Nope, not tested it yet.

**Comment by rade** [comment]
I was actually more thinking of the boilerplate reduction than the latency. Because of vendoring it's hard to tell the actual difference at a glance.

**Comment by tomwilkie** [comment]
Take a look at the second commit.  We currently haven;t removed the origin code http (as its needed for migration), but will do.  I just tidied it up for now.

**Comment by tomwilkie** [comment]
Had to update built-tools and fix the generation of bindata.go to get lint to pass.

**Comment by tomwilkie** [comment]
@jml this is good to go now.  Anecdotally query performance is much improved (locally).

File: distributor/http.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	var req remote.WriteRequest
+	ctx, abort := util.ParseProtoRequest(w, r, &req, true)
+	if abort {
+		return
+	}
+
+	_, err := d.Push(ctx, &req)
+	if err != nil {
+		switch e := err.(type) {
+		case IngesterError:
```
Comment (jml) [inline distributor/http.go]: AIUI, this is only going to be relevant if we're also using the HTTP ingester client, and that with gRPC, we will no longer be able to distinguish between user errors and server errors. Is there a way of addressing this? It's a bit of a regression.

File: cmd/cortex/main.go [hunk 2]
```diff
@@ -84,30 +88,38 @@ func main() {
...
 	flag.IntVar(&cfg.distributorConfig.MinReadSuccesses, "distributor.min-read-successes", 2, "The minimum number of ingesters from which a read must succeed.")
 	flag.DurationVar(&cfg.distributorConfig.HeartbeatTimeout, "distributor.heartbeat-timeout", time.Minute, "The heartbeat timeout after which ingesters are skipped for reads/writes.")
+	flag.DurationVar(&cfg.distributorConfig.RemoteTimeout, "distributor.remote-timeout", 5*time.Second, "Timeout for downstream ingesters.")
+
 	flag.StringVar(&cfg.rulerConfig.ConfigsAPIURL, "ruler.configs.url", "", "URL of configs API server.")
 	flag.StringVar(&cfg.rulerConfig.UserID, "ruler.userID", "", "Weave Cloud org to run rules for")
 	flag.DurationVar(&cfg.rulerConfig.EvaluationInterval, "ruler.evaluation-interval", 15*time.Second, "How frequently to evaluate rules")
-	flag.BoolVar(&cfg.logSuccess, "log.success", false, "Log successful requests")
-	flag.BoolVar(&cfg.watchDynamo, "watch-dynamo", false, "Periodically collect DynamoDB provisioned throughput.")
+
```
Comment (tomwilkie) [inline cmd/cortex/main.go]: For my own sanity, I grouped the flags by common, ingester, distributor etc

File: cmd/cortex/main.go [hunk 3]
```diff
@@ -139,21 +151,34 @@ func main() {
...
+		grpcServer := grpc.NewServer(
+			grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(
+				cortex_grpc_middleware.ServerInstrumentInterceptor(requestDuration),
+				cortex_grpc_middleware.ServerLoggingInterceptor(cfg.logSuccess),
+				cortex_grpc_middleware.ServerUserHeaderInterceptor,
+			)),
+		)
+		cortex.RegisterIngesterServer(grpcServer, ing)
+		go grpcServer.Serve(lis)
+		defer grpcServer.Stop()
```
Comment (tomwilkie) [inline cmd/cortex/main.go]: Yeah.  I think we'll end up breaking this into a bunch of binaries / images next.

File: distributor/distributor.go [hunk 4]
```diff
@@ -158,13 +173,14 @@ type sampleTracker struct {
 	succeeded  int32
 }
 
-// Append implements SampleAppender.
-func (d *Distributor) Append(ctx context.Context, samples []*model.Sample) error {
+// Push Implements cortex.IngesterServer
```
Comment (tomwilkie) [inline distributor/distributor.go]: Yeah, thats the idea.  Just this function really.  We do use the IngesterClient internally.

File: ruler/ruler.go [hunk 5]
```diff
@@ -151,12 +151,14 @@ func (r *Ruler) getConfig(userID string) (*config.Config, error) {
 
 // appenderAdapter adapts cortex.SampleAppender to prometheus.SampleAppender
 type appenderAdapter struct {
-	appender cortex.SampleAppender
-	ctx      context.Context
+	distributor *distributor.Distributor
```
Comment (tomwilkie) [inline ruler/ruler.go]: Not fully - its only implements the Push method.  I thought the coupling was a bit strong here, but there is no other interface just for Push, unless we invent one.

File: distributor/http.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
+package distributor
```
Comment (tomwilkie) [inline distributor/http.go]: Don't mind either way.  The ingester_client.go and ingester/http.go are going to be deleted in the next PR.

File: util/compat.go [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+		for _, s := range ts.Samples {
+			ss.Values = append(ss.Values, model.SamplePair{
+				Value:     model.SampleValue(s.Value),
+				Timestamp: model.Time(s.TimestampMs),
+			})
+		}
+		m = append(m, &ss)
+	}
+
+	return m
```
Comment (tomwilkie) [inline util/compat.go]: Done.

**Comment by tomwilkie** [comment]
@jml  PTAL

File: distributor/http.go (hunk 1 above)
Comment (jml) [inline distributor/http.go]: Can you please file a bug for follow-up?

File: util/middleware/logging.go [hunk 8]
```diff
@@ -15,8 +15,10 @@ func ServerLoggingInterceptor(logSuccess bool) grpc.UnaryServerInterceptor {
 	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
 		begin := time.Now()
 		resp, err := handler(ctx, req)
-		if logSuccess || err != nil {
+		if logSuccess && err != nil {
```
Comment (jml) [inline util/middleware/logging.go]: That should be `err == nil`, I think. Probably overall clearer to do:

```go
if err != nil {
  ...
} else if logSuccess {
  ...
}
```

**Comment by jml** [comment]
Thanks. Minor tweaks.

**Pull request merged**